distributed/utils_test.py

```diff
@@ -1793,6 +1795,11 @@ def check_instances():
 
     _global_clients.clear()
 
+    wait_profiler()
+    gc.collect()
```

Comment:
I used this myself already but this is actually not safe. The `wait_profiler` polls for the thread to not be running. If we rely on ordinary ref-count based object collection, this is sufficient since as soon as the profiler thread pauses, the ref-count based collection cleans up all references and we're good.

Most of the `TaskState`s are actually part of a densely connected, self referencing data structure. these self referencing cycles are not necessarily a problem since the gc.collect can detect these, break them and clean up. However, the background thread may actually already be running again after leaving `wait_profiler` since we're only polling and are not actually stopping the thread.

Reply:
We'd rather need something like

```python
with no_profiler():
    # Some magic (e.g. a lock) ensures that the profile thread cannot watch while this ctx manager is held
    gc.collect()
```

Reply:
Maybe this?

```python
from distributed import profile

with profile.lock:
    gc.collect()
```

```python
# profile.py

lock = threading.Lock()

def _watch(...):
    ...
    with lock:
        frame = sys._current_frames()[thread_id]
```

Reply:
(it seems simpler to reference a lock directly rather than make a context manager)

Reply:
yes, I was typing the contextmanager first and while typing realized this is not occult magic but simply a lock :)